### PR TITLE
remove deserialize mode MODULE_POSTWORK

### LINKED ISSUE #18343: Crash on master with CategoricalArrays
I'm getting a crash both with latest master and with 0.4.6 after loading CategoricalArrays.jl and calling `gc()`. You can get the code from the `nl/crash` branch.
```               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.6.0-dev.456 (2016-09-03 13:57 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit a9140fd* (0 days old master)
|__/                   |  x86_64-redhat-linux

julia> using CategoricalArrays
INFO: Recompiling stale cache file /home/milan/.julia/lib/v0.6/CategoricalArrays.ji for module CategoricalArrays.
julia> gc()
GC error (probable corruption) :
Allocations: 2519259 (Pool: 2518041; Big: 1218); GC: 2
<?#0x7f85f4112870::<circular reference @-1>>

signal (6): Aborted
while loading no file, in expression starting on line 0
gsignal at /lib64/libc.so.6 (unknown line)
abort at /lib64/libc.so.6 (unknown line)
push_root at /home/milan/Dev/julia/src/gc.c:1478
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
gc_mark_module at /home/milan/Dev/julia/src/gc.c:1218
push_root at /home/milan/Dev/julia/src/gc.c:1429
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1368
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
gc_mark_module at /home/milan/Dev/julia/src/gc.c:1218
push_root at /home/milan/Dev/julia/src/gc.c:1429
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
push_root at /home/milan/Dev/julia/src/gc.c:1466
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
gc_mark_module at /home/milan/Dev/julia/src/gc.c:1218
push_root at /home/milan/Dev/julia/src/gc.c:1429
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
gc_mark_module at /home/milan/Dev/julia/src/gc.c:1218
push_root at /home/milan/Dev/julia/src/gc.c:1429
gc_push_root at /home/milan/Dev/julia/src/gc.c:1189 [inlined]
pre_mark at /home/milan/Dev/julia/src/gc.c:1515
_jl_gc_collect at /home/milan/Dev/julia/src/gc.c:1709 [inlined]
jl_gc_collect at /home/milan/Dev/julia/src/gc.c:1865
gc at ./base.jl:145
unknown function (ip: 0x7f85e315de7f)
jl_call_method_internal at /home/milan/Dev/julia/src/julia_internal.h:189 [inlined]
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:1931
do_call at /home/milan/Dev/julia/src/interpreter.c:66
eval at /home/milan/Dev/julia/src/interpreter.c:205
jl_toplevel_eval_flex at /home/milan/Dev/julia/src/toplevel.c:619 [inlined]
jl_toplevel_eval at /home/milan/Dev/julia/src/toplevel.c:641
jl_toplevel_eval_in_warn at /home/milan/Dev/julia/src/builtins.c:590
eval at ./boot.jl:234
unknown function (ip: 0x7f87f6301cdf)
jl_call_method_internal at /home/milan/Dev/julia/src/julia_internal.h:189 [inlined]
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:1931
eval_user_input at ./REPL.jl:66
unknown function (ip: 0x7f85e3156a06)
jl_call_method_internal at /home/milan/Dev/julia/src/julia_internal.h:189 [inlined]
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:1931
macro expansion at ./REPL.jl:97 [inlined]
#3 at ./event.jl:68
unknown function (ip: 0x7f85e314df2f)
jl_call_method_internal at /home/milan/Dev/julia/src/julia_internal.h:189 [inlined]
jl_apply_generic at /home/milan/Dev/julia/src/gf.c:1931
jl_apply at /home/milan/Dev/julia/src/julia.h:1364 [inlined]
start_task at /home/milan/Dev/julia/src/task.c:259
unknown function (ip: 0xffffffffffffffff)
Allocations: 2519259 (Pool: 2518041; Big: 1218); GC: 2
Aborted (core dumped)

```

Is this a duplicate of https://github.com/JuliaLang/julia/issues/18299?

### PR DESCRIPTION
**Pull request by vtjnash**
it seems to be unnecessary

fix #18343

### CODE REVIEW

**Comment by vtjnash** [comment]
CI failure was unrelated:
	From worker 3:	     * pkg                  Exception running test pkg :
On worker 3:
LoadError: GitError(Code:ERROR, Class:OS, Failed to receive response: The server returned an invalid or unrecognized response
)

File: src/dump.c [hunk 1]
```diff
@@ -1792,16 +1758,58 @@ static jl_value_t *jl_deserialize_value_(jl_serializer_state *s, jl_value_t *vta
...
 
-static void jl_deserialize_lambdas_from_mod(jl_serializer_state *s)
+typedef struct _linkedlist_t {
+    struct _linkedlist_t *next;
+    struct {
+        jl_method_t *meth;
+        jl_tupletype_t *simpletype;
+    } def[100];
+    size_t count;
+} linkedlist_t;
```
Comment (tkelman) [inline src/dump.c]: maybe this file should just be c++ then?

File: src/dump.c [hunk 2]
```diff
@@ -1714,15 +1686,16 @@ static jl_value_t *jl_deserialize_value_(jl_serializer_state *s, jl_value_t *vta
     else if (vtag == (jl_value_t*)jl_datatype_type || vtag == (jl_value_t*)SmallDataType_tag) {
         int32_t sz = (vtag == (jl_value_t*)SmallDataType_tag ? read_uint8(s->s) : read_int32(s->s));
         jl_value_t *v = jl_gc_alloc(ptls, sz, NULL);
+        jl_set_typeof(v, (void*)(intptr_t)0x40);
```
Comment (tkelman) [inline src/dump.c]: does 0x40 have a define'd name?

**Comment by tkelman** [comment]
what was MODE_MODULE_POSTWORK supposed to be for then?

**Comment by vtjnash** [comment]
trying to be more efficient (another of my recent patches mostly resolved that though), and to avoid writing the linked list implementation

File: src/dump.c (hunk 2 above)
Comment (tkelman) [inline src/dump.c]: yes or no?

Comment (vtjnash) [inline src/dump.c]: no, it's just a random garbage value (0x4, since it's the 4th one added here), like setting it to NULL, but more likely to fail

Comment (tkelman) [inline src/dump.c]: that should be a comment

Comment (tkelman) [inline src/dump.c]: add a comment here

**Comment by tkelman** [comment]
being backported in #18869

**Pull request merged**